File: IAIso-v5.0/core/iaiso-swift/Sources/IAIsoMiddleware/Cohere.swift
import Foundation
import IAIsoCore

/// Cohere chat middleware.
public enum Cohere {

    public protocol Client {
        func chat(_ params: [String: Any]) throws -> Response
    }

    public struct BilledUnits: Sendable, Equatable {
        public let inputTokens: Int64
        public let outputTokens: Int64
        public init(inputTokens: Int64 = 0, outputTokens: Int64 = 0) {
            self.inputTokens = inputTokens
            self.outputTokens = outputTokens
        }
    }

    public struct Meta: Sendable, Equatable {
        public let tokens: BilledUnits?
        public let billedUnits: BilledUnits?
        public init(tokens: BilledUnits? = nil, billedUnits: BilledUnits? = nil) {
            self.tokens = tokens
            self.billedUnits = billedUnits
        }
    }

    public struct ToolCall: Sendable, Equatable {
        public let name: String
        public init(name: String) { self.name = name }
    }

    public struct Response: Sendable {
        public let model: String
        public let meta: Meta
        public let toolCalls: [ToolCall]
        public init(model: String, meta: Meta, toolCalls: [ToolCall] = []) {
            self.model = model
            self.meta = meta
            self.toolCalls = toolCalls
        }
    }

    public struct Options: Sendable {
        public var raiseOnEscalation: Bool
        public init(raiseOnEscalation: Bool = false) {
            self.raiseOnEscalation = raiseOnEscalation
        }
        public static let defaults = Options()
    }

    public final class BoundedClient {
        private let raw: Client
        private let execution: BoundedExecution
        private let opts: Options

        public init(raw: Client, execution: BoundedExecution, options: Options = .defaults) {
            self.raw = raw
            self.execution = execution
            self.opts = options
        }

        public func chat(_ params: [String: Any]) throws -> Response {
            let pre = execution.check()
            if pre == .locked { throw MiddlewareError.locked }
            if pre == .escalated, opts.raiseOnEscalation {
                throw MiddlewareError.escalationRaised
            }
            let resp: Response
            do {
                resp = try raw.chat(params)
            } catch {
                throw MiddlewareError.provider(
                    error.localizedDescription, underlying: error)
            }
            let units = resp.meta.tokens ?? resp.meta.billedUnits
            let tokens = (units != nil)
                ? (units!.inputTokens + units!.outputTokens) : 0
            let toolCalls = Int64(resp.toolCalls.count)
            let model = resp.model.isEmpty ? "unknown" : resp.model
            execution.recordStep(StepInput(
                tokens: tokens,
                toolCalls: toolCalls,
                tag: "cohere.chat:\(model)"))
            return resp
        }
    }
}

File: IAIso-v5.0/core/iaiso-rust/crates/middleware/src/cohere.rs
//! Cohere API wrapper.

use crate::MiddlewareError;
use iaiso_core::{BoundedExecution, StepInput, StepOutcome};
use serde::Deserialize;
use serde_json::Value;

#[derive(Debug, Clone, Deserialize, Default)]
#[serde(rename_all = "camelCase")]
pub struct BilledUnits {
    #[serde(default)]
    pub input_tokens: u64,
    #[serde(default)]
    pub output_tokens: u64,
}

#[derive(Debug, Clone, Deserialize, Default)]
pub struct Meta {
    #[serde(default)]
    pub tokens: Option<BilledUnits>,
    #[serde(default, rename = "billedUnits")]
    pub billed_units: Option<BilledUnits>,
}

#[derive(Debug, Clone, Deserialize)]
pub struct ToolCall {
    pub name: String,
}

#[derive(Debug, Clone, Deserialize)]
pub struct Response {
    #[serde(default)]
    pub model: String,
    #[serde(default)]
    pub meta: Meta,
    #[serde(default, rename = "toolCalls")]
    pub tool_calls: Vec<ToolCall>,
}

pub trait Client: Send + Sync {
    fn chat(&self, params: &Value) -> Result<Response, String>;
}

#[derive(Debug, Clone, Copy, Default)]
pub struct Options {
    pub raise_on_escalation: bool,
}

pub struct BoundedClient<'a, C: Client> {
    raw: C,
    execution: &'a BoundedExecution,
    opts: Options,
}

impl<'a, C: Client> BoundedClient<'a, C> {
    pub fn new(raw: C, execution: &'a BoundedExecution, opts: Options) -> Self {
        Self { raw, execution, opts }
    }

    pub fn chat(&self, params: &Value) -> Result<Response, MiddlewareError> {
        match self.execution.check() {
            StepOutcome::Locked => return Err(MiddlewareError::Locked),
            StepOutcome::Escalated if self.opts.raise_on_escalation => {
                return Err(MiddlewareError::EscalationRaised)
            }
            _ => {}
        }
        let resp = self.raw.chat(params).map_err(MiddlewareError::Provider)?;
        let bucket = resp
            .meta
            .tokens
            .as_ref()
            .or(resp.meta.billed_units.as_ref());
        let tokens = bucket
            .map(|b| b.input_tokens + b.output_tokens)
            .unwrap_or(0);
        let tool_calls = resp.tool_calls.len() as u64;
        let model = if resp.model.is_empty() {
            "unknown".to_string()
        } else {
            resp.model.clone()
        };
        let _ = self.execution.record_step(StepInput {
            tokens,
            tool_calls,
            tag: Some(format!("cohere.chat:{}", model)),
            ..Default::default()
        });
        Ok(resp)
    }
}
